clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -1136,56 +1153,128 @@ public void assign(Collection<TopicPartition> partitions) {
+     * @throws java.lang.IllegalArgumentException if the timeout value is negative
+     * @throws java.lang.IllegalStateException if the consumer is not subscribed to any topics or manually assigned any
+     *             partitions to consume from
+     * @throws java.lang.ArithmeticException if the timeout is greater than {@link Long#MAX_VALUE} milliseconds.
      */
-    private Map<TopicPartition, List<ConsumerRecord<K, V>>> pollOnce(long timeout) {
-        client.maybeTriggerWakeup();
+    @Override
+    public ConsumerRecords<K, V> poll(final Duration timeout) {
+        if (timeout.isNegative()) throw new IllegalArgumentException("Timeout must not be negative");
+
+        acquireAndEnsureOpen();
+        try {
+            if (this.subscriptions.hasNoSubscriptionOrUserAssignment())
+                throw new IllegalStateException("Consumer is not subscribed to any topics or assigned any partitions");
+
+            // poll for new data until the timeout expires
+            final Instant start = Instant.ofEpochMilli(time.milliseconds());
+            do {
+
+                client.maybeTriggerWakeup();
 
-        long startMs = time.milliseconds();
-        coordinator.poll(startMs, timeout);
+                if (!internalUpdateAssignmentMetadataIfNeeded(remainingTimeAtLeastZero(start, timeout))) {
+                    return ConsumerRecords.empty();
+                }
 
-        // Lookup positions of assigned partitions
-        boolean hasAllFetchPositions = updateFetchPositions();
+                final Map<TopicPartition, List<ConsumerRecord<K, V>>> records =
+                    pollForFetches(remainingMsAtLeastZero(start.toEpochMilli(), timeout));
 
+                if (!records.isEmpty()) {
+                    // before returning the fetched records, we can send off the next round of fetches
+                    // and avoid block waiting for their responses to enable pipelining while the user
+                    // is handling the fetched records.
+                    //
+                    // NOTE: since the consumed position has already been updated, we must not allow
+                    // wakeups or any other errors to be triggered prior to returning the fetched records.
+                    if (fetcher.sendFetches() > 0 || client.hasPendingRequests()) {
+                        client.pollNoWakeup();
+                    }
+
+                    return this.interceptors.onConsume(new ConsumerRecords<>(records));
+                }
+
+            } while (!remainingTimeAtLeastZero(start, timeout).isZero());
+
+            return ConsumerRecords.empty();
+        } finally {
+            release();
+        }
+    }
+
+
+    /**
+     * Visible for testing
+     */
+    @SuppressWarnings("BooleanMethodIsAlwaysInverted") // because false => timed out, in which case we return early or throw.
+    boolean internalUpdateAssignmentMetadataIfNeeded(final Duration timeout) {
```

Comment:
I called it "internal" because this method isn't protected by the light lock. It seemed like a risk that someone might want to expose this method in the future, and I wanted them to think twice about just making it public.

Alternatively, I can go ahead and (re)acquire the lock in here, or just change the name to "updateAssignmentMetadataIfNeeded" and leave the implementation as-is.

I'm happy to do whatever you prefer.